### DISCUSSION TITLE
Insert position and scale factors seems wrong

### DISCUSSION THREAD

**Discussion by mgnslndh** [DomCR/ACadSharp#175] [Q&A]
I have problems interpreting the values of a block insert from a DWG.

When I use ACadSharp, QCAD and DWGSee to inspect the properties of the insert I get different values. 

|Tool|Insert X|Insert Y|Insert Z|Scale X|Scale Y|Scale Z|Rotation
|-----|------|------|------|------|----|------|-----|
|ACadSharp|-26904.778108801958|13290.592388701658|0|-1|1|-1|Π radians|
|QAD|26904.77810880|13290.59238870|N/A|-1|-1|N/A|0|
|DWGSee|26904.778|13290.592|0.000|-1|1|-1|180 degrees|
|Teigha 24.8|26904.77810880196|13290.592388701658|0|-1|1|-1|3.141592653589793|

Is there a bug or is there something I don't understand to interpret these values correctly?

The handle of the entity is 0x38730 and can be seen in both screenshots for ACadSharp and QCAD. It is not possible to display this handle in DWGSee. 

I'm thinking maybe ACadSharp is reading the values correctly but the other tools are converting the values before using/showing them. 

**QAD**
![image](https://github.com/DomCR/ACadSharp/assets/1139118/1f74a01c-22fe-47df-8847-a2cf109cd414)

**DWGSee**
![image](https://github.com/DomCR/ACadSharp/assets/1139118/10bc0825-9a3c-4e4f-94fa-3c74091c206b)

![image](https://github.com/DomCR/ACadSharp/assets/1139118/89c01541-b30a-484f-82cd-7f5d122884dd)

**Comment by mgnslndh** [comment]
If I save the file from QCAD without making changes and read it in ACadSharp I get the same values as from QCAD. Seems like the tools are making some conversions.

![image](https://github.com/DomCR/ACadSharp/assets/1139118/81abc189-f725-4c87-a3f5-5b2fcd381991)

What I do not understand is if the original file was corrupted or if the tools just make some conversion that are not obvious to me?

**Comment by DomCR** [comment]
Hi @mgnslndh 

**Insert X**: not sure why is the only one negative, seems odd that does not match with any other framework, did you try to open the file using Autocad and check these values? 

**Rotation**: I spoted an error in the `ACadSharp` when reading the insert that was fixed in this commit https://github.com/DomCR/ACadSharp/commit/56b2eff3c9df8c5ab2212f95fba6968843908fe9 so the rotation in the table should be 180 degrees instead of PI if you pull the latest version this issue should be fixed.

One possible solution it would be to export the file in a DXF and check the values manually, if you have the handle is quite easy, you can find the DXF codes in this [link](https://help.autodesk.com/view/OARX/2021/ENU/?guid=GUID-235B22E0-A567-4CF6-92D3-38A2306D73F3), hope this helps.

**Comment by mgnslndh** [comment]
Thanks for the response @DomCR! 

I do not have access to AutoCAD at the moment so I can not try that. But all programs I have that can open a DWG show me the correct value for the X coordinate and the drawing looks OK on screen.

If I save the drawing in QCAD as DWG or DXF the value will be interpreted correctly by ACadSharp.

I made another test, this time using Teigha 24.8 and the results looks very much like ACadSharp except for the X coordinate which is positive as I would expect it to be. Can there be an issue regarding the sign bit when reading the X coordinate in ACadSharp?

**Comment by mgnslndh** [comment]
@DomCR if you want to troubleshoot the issue I can email you the DWG file.

**Reply by DomCR** [reply]
If you email the file I'll take a look to see if there is an issue reading the block.

**Reply by mgnslndh** [reply]
Thanks @DomCR! I've sent you an email with the DWG. Let me know if there is anything I can do.

**Comment by DomCR** [comment]
![image](https://github.com/DomCR/ACadSharp/assets/46492487/41817a0f-f98d-449d-829c-f30365d93f7b)

The block structure is the following; the current `Insert` entity is an instance of the `BLOCK_RECORD` named **NV24MS** (Property Block in the image) which is contained in the `BLOCK_RECORD` named **pos 304_u** (Property Owner in the image) if you go to the block record **pos 304_u** you will see all the entities that create the block, one of them will be an insert that is referenced to **NV24MS**.

Now the position in of the block **NV24MS** in Autocad seems to be (26904.778108801958, 13290.592388701658) which it does not match with ACadSharp, one thing that I've noticed is that the Normal is flipped in the Z axis, I couldn't find a way to check the normal value for an insert in Autocad, but maybe it does some sort of transformation and corrects both values to simplify the reading. Can you check the value of the Normal in the other frameworks?

**Reply by mgnslndh** [reply]
I can confirm that the normal is 0,0,-1 using Teigha but the other tools do not show the normal in the entity properties. I have also thought about some kind of translation going on but I do not know if that is actually the case. If you think about the normal as from what direction you are looking at the entity it would be as if the camera was behind the screen looking back at the screen and that would flip the x-axis. Not sure that is how it works though. Most other entities in the DWG have a normal of 0,0,1.

**Comment by mgnslndh** [comment]
I found out that QAD have made some kind of handling of inverted extrusion vectors (normal in ACadSharp):

>Files imported from other CAD applications might contain blocks that are mirrored if the block reference uses an inverted extrusion direction. [Source](https://qcad.org/bugtracker/index.php?do=details&task_id=1550)

I've not found the commits or source code that fixes this issue but it is reported as fixed. QCAD will load the DWG and the X axis is mirrored in relation to what we can read with ACadSharp.

I also found this discussion about how to handle entities with inverted extrusion vectors but i'm not sure it is helpful. They do talk about flipping the geometry:

> If the block is asymetric wouldn't modifying the the extrusion vector from -1 to 1 have the affect of flipping the geometry? [Source](https://www.cadtutor.net/forum/topic/37869-mirrored-blocks-and-extrusions/)
